lib/rails_multitenant/global_context_registry.rb

```diff
@@ -26,21 +26,80 @@ module Current
 
       module ClassMethods
         def current
-          GlobalContextRegistry.fetch(current_registry_obj) { new }
+          GlobalContextRegistry.fetch(current_registry_obj) { __current_default }
+        end
+
+        def current=(object)
+          raise "#{object} is not a #{self}" if object.present? && !object.is_a?(self)
+          GlobalContextRegistry.set(current_registry_obj, object)
+          __clear_dependents!
+        end
+
+        def current?
+          GlobalContextRegistry.get(current_registry_obj).present?
+        end
+
+        def current!
+          current || raise("No current #{name} set")
         end
 
         def clear_current!
           GlobalContextRegistry.delete(current_registry_obj)
         end
 
+        def as_current(object)
+          old_object = current if current?
+          self.current = object
+          yield
+        ensure
+          self.current = old_object
+        end
+
         include RegistryDependentOn
 
+        def provide_default(provider = nil, &block)
+          @default_provider = provider ? provider.to_proc : block
+        end
+
         private
 
         def current_registry_obj
-          key_class = respond_to?(:base_class) ? base_class : self
-          "#{key_class.name.underscore}_obj".to_sym
+          return @current_registry_obj if @current_registry_obj
+
+          @current_registry_obj = "#{__key_class.name.underscore}_obj".to_sym
         end
+
+        def current_registry_default_provider
+          "#{__key_class.name.underscore}_default_provider".to_sym
+        end
+
+        def __current_default
+          if @default_provider
```

Comment:
Could initialize `@default_provider` to `Proc.new {}` to avoid this conditional.